### PR TITLE
added messageRecall forwardMessage implementation

### PR DESCRIPTION
**Pull request by guo40020**

### CODE REVIEW

**Comment by bung87** [comment]
messageRecall is action

File: src/puppet-whatsapp.ts [hunk 1]
```diff
@@ -477,7 +477,7 @@ class PuppetWhatsapp extends PUPPET.Puppet {
   }
 
   /**
-   * recall the message
+   * check if the message is recalled
```
Comment (bung87) [inline src/puppet-whatsapp.ts:480]: messageRecall is action

**Review by bung87 — CHANGES_REQUESTED** [review]

File: src/puppet-whatsapp.ts [hunk 2]
```diff
@@ -600,8 +606,10 @@ class PuppetWhatsapp extends PUPPET.Puppet {
     log.verbose('PuppetWhatsApp', 'messageForward(%s, %s)', conversationId, messageId)
     const msg = this.messageStore[messageId]
     if (!msg) {
-      log.error('')
+      log.error('Message %s not found', messageId)
+      throw new Error('Message not found')
```
Comment (su-chang) [inline src/puppet-whatsapp.ts:610]: It's better to add `messageId` to error message.

File: src/puppet-whatsapp.ts [hunk 3]
```diff
@@ -486,9 +486,15 @@ class PuppetWhatsapp extends PUPPET.Puppet {
     const msg = this.messageStore[messageId]
     if (!msg) {
       log.error('Message %s not found', messageId)
+      throw new Error('Message not found')
+    }
+
+    try {
+      await msg.delete(true)
```
Comment (su-chang) [inline src/puppet-whatsapp.ts:493]: Question about `msg.delete(true)`, it will return void or not?

Comment (guo40020) [reply src/puppet-whatsapp.ts:493]: void

File: src/puppet-whatsapp.ts [hunk 4]
```diff
@@ -600,8 +606,10 @@ class PuppetWhatsapp extends PUPPET.Puppet {
     log.verbose('PuppetWhatsApp', 'messageForward(%s, %s)', conversationId, messageId)
     const msg = this.messageStore[messageId]
     if (!msg) {
-      log.error('')
+      log.error('Message %s not found', messageId)
+      throw new Error('Message not found')
     }
+    await msg.forward(conversationId)
```
Comment (su-chang) [inline src/puppet-whatsapp.ts:612]: messageForward (conversationId: string, messageId: string): Promise<void | string>

`string`: new message id

**Comment by bung87** [comment]
Note here:  

in whatsapp-web.js internal , it calls `sendDeleteMsgs` which is undocumented.  

```
async delete(everyone) {
        await this.client.pupPage.evaluate((msgId, everyone) => {
            let msg = window.Store.Msg.get(msgId);

            if (everyone && msg.id.fromMe && msg.canRevoke()) {
                return window.Store.Cmd.sendRevokeMsgs(msg.chat, [msg], true);
            }

            return window.Store.Cmd.sendDeleteMsgs(msg.chat, [msg], true);
        }, this.id._serialized, everyone);
    }
```

File: src/puppet-whatsapp.ts (hunk 4 above)
Comment (guo40020) [reply src/puppet-whatsapp.ts:612]: so we are changing the signature?

File: src/puppet-whatsapp.ts (hunk 2 above)
Comment (guo40020) [reply src/puppet-whatsapp.ts]: done

**Comment by bung87** [comment]
refers to https://github.com/wechaty/puppet-whatsapp/issues/26#issuecomment-1011839732

File: src/puppet-whatsapp.ts (hunk 4 above)
Comment (guo40020) [reply src/puppet-whatsapp.ts:612]: done

**Pull request merged**